Add policy check for SharePoint 3.2 when using service principal and update SharePoint 4.2 rego for deprecation

mitchelbaker-cisa opened the pull request:
# <!-- Use the title to describe PR changes in the imperative mood --> #
  <!-- Remember this title will end up as the merge commit subject -->

## 🗣 Description ##

- Removed `input.OneDrive_PnP_Flag == false/true` statements and Not-Implemented test in `SharePointConfig.rego`. The `input.OneDrive_PnP_Flag` is still used in the SharePoint unit tests to indicate spo/pnp variants
- Removed unnecessary fields from `sharepoint.spo.testplan.yaml` to get rid of warnings; removed deprecated parameter ("Detailed"=$true) from Get-SPOSite and Get-PnPTenantSite in SharePoint provider
- Expanded unit test coverage for SharePoint 3.2 to cover service principal cases in `SharepointConfig_03_test.rego`; expanded functional test coverage for SharePoint 3.2 in the `sharepoint.pnp.testplan.yaml`

- Logic for SharePoint 3.1 fixed to account for negative integer value, -1, when the Anyone links checkbox is not set. The policy now displays a fail for this case
- Set SharePoint 4.2 to not-implemented in `SharepoingConfig.rego`, updated its unit/functional tests

## 💭 Motivation and context ##

<!-- Why is this change required? -->
<!-- What problem does this change solve? How did you solve it? -->
<!-- Mention any related issue(s) here using appropriate keywords such -->
<!-- as "closes" or "resolves" to auto-close them on merge. -->

Closes #1221 
Closes #1268 
Closes #1220 
Closes #1324 
closes #622 

## 🧪 Testing ##

<!-- How did you test your changes? How could someone else test this PR? -->
<!-- Include details of your testing environment, and the tests you ran to -->
<!-- see how your change affects other areas of the code, etc. -->

Checkout the branch locally:
```
git fetch
git checkout 1221-sharepoint-3.2-incorrect-NA
```

Check if all unit tests pass: 
```
.\Testing\RunUnitTests.ps1 -p sharepoint
```

Check if all functional tests pass against SharePoint, use both spo/pnp variants.

<!--
## 📷 Screenshots (if appropriate) ##

Uncomment this section if a screenshot is needed.

-->

## ✅ Pre-approval checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- Draft PRs may have one or more unchecked boxes. -->
<!-- If you're unsure about any of these, don't hesitate to ask. -->
<!-- We're here to help! -->

- [x] This PR has an informative and human-readable title.
- [x] PR targets the correct parent branch (e.g., main or release-name) for merge.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] Changes are sized such that they do not touch excessive number of files.
- [x] *All* future TODOs are captured in issues, which are referenced in code comments.
- [x] These code changes follow the ScubaGear [content style guide](https://github.com/cisagov/ScubaGear/blob/main/CONTENTSTYLEGUIDE.md).
- [x] Related issues these changes resolve are linked preferably via [closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).
- [x] All relevant type-of-change labels added.
- [x] All relevant project fields are set.
- [x] All relevant repo and/or project documentation updated to reflect these changes.
- [x] Unit tests added/updated to cover PowerShell and Rego changes.
- [x] Functional tests added/updated to cover PowerShell and Rego changes.
- [x] All relevant functional tests passed.
- [x] All automated checks (e.g., linting, static analysis, unit/smoke tests) passed.

## ✅ Pre-merge checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- These boxes should remain unchecked until the pull request has been -->
<!-- approved. -->

- [x] PR passed smoke test check.
- [ ] Feature branch has been rebased against changes from parent branch, as needed

  Use `Rebase branch` button below or use [this](https://www.digitalocean.com/community/tutorials/how-to-rebase-and-update-a-pull-request) reference to rebase from the command line.
- [x] Resolved all merge conflicts on branch
- [ ] Notified merge coordinator that PR is ready for merge via comment mention

## ✅ Post-merge checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- These boxes should remain unchecked until the pull request has been -->
<!-- approved. This section is for the merge coordinator to complete. -->
- [ ] Feature branch deleted after merge to clean up repository.
- [ ] Verified that all checks pass on parent branch (e.g., main or release-name) after merge.

tkol2022 commented:
It is not clear why the test_File_Folder_AnonymousLinkType_Correct should produce a Pass? The SharingCapability is explicitly set to 2 (Anyone) within the unit test, but the other fields that are checked by the policy FileAnonymousLinkType & FolderAnonymousLinkType are not set in the unit test. Since they are already setup to the compliant value of 1 in the SharepointBaseConfig.rego file, the policy passes, but as a developer when I read the unit test this is not clear and I think this will make it hard to maintain these tests and know what they doing. I am wondering if we should explicitly set the values related to the policy and scenario we are testing within each unit test?

![image](https://github.com/user-attachments/assets/85b3d18b-959a-47d7-8256-9e36ac481dbd)

![image](https://github.com/user-attachments/assets/2de2f4ad-c204-4c09-b2e4-1c3be0d01b7d)

tkol2022 commented:
I don't see the difference between these two tests?
Also, what does the "UsingServicePrincipal" in the test name mean? I don't see anything in the test JSON that indicates that a service principal was used.

![image](https://github.com/user-attachments/assets/f6b06590-fc0b-4cff-8a5d-a814e55f0569)

![image](https://github.com/user-attachments/assets/71a79eab-8b0e-4356-97f1-287198c8286c)

I don't see the difference between the two tests below either?

![image](https://github.com/user-attachments/assets/8f328232-7fec-4fd9-b296-afe666ccb4fa)

![image](https://github.com/user-attachments/assets/0a6407e0-642d-4765-9598-1d973c66f493)

tkol2022 reviewed (changes requested):
Reviewed the code updates, unit tests and functional tests. Will leverage the support of another team member to perform hands-on testing.

I had a couple of minor questions related to the consistency and maintainability of the unit tests. Other than that, looks good. Nice work.

mitchelbaker-cisa commented:
> Reviewed the code updates, unit tests and functional tests. Will leverage the support of another team member to perform hands-on testing.
> 
> I had a couple of minor questions related to the consistency and maintainability of the unit tests. Other than that, looks good. Nice work.

Thanks for the review. I addressed your feedback in https://github.com/cisagov/ScubaGear/pull/1309/commits/7bc097513451e008800a6ed014a4eaa4dd1f3db1

Sloane4 commented on PowerShell/ScubaGear/Testing/Unit/Rego/Sharepoint/SharepointConfig_03_test.rego:
```diff
@@ -153,73 +153,146 @@ test_SharingCapability_NewExistingGuests_NotApplicable_V2 if {
+    # Set PnP flag to true denoting use of service principal
     Output := sharepoint.tests with input.SPO_tenant as [Tenant]
-                                with input.OneDrive_PnP_Flag as true
+        with input.OneDrive_PnP_Flag as true
```
I thought we were removing this?

mitchelbaker-cisa replied:
It's only removed in the SharePoint.rego file. When running Invoke-SCuBA with the AppId parameter this sets `$PnPFlag = $true` in Orchestrator.ps1.

![Screenshot (230)](https://github.com/user-attachments/assets/d6d742b0-3c25-473b-bdef-28e8fbe92a4c)

It's then passed into the SharePoint provider to conditionally run either Get-PnPTenant/Get-PnpTenantSite or Get-SPOTenant/Get-SPOSite commandlets. 

![Screenshot (232)](https://github.com/user-attachments/assets/70e2deb1-1459-4be5-857e-9a29c3611883)

The purpose of keeping `input.OneDrive_PnP_Flag as true` in the unit tests is to mock running Invoke-SCuBA via service principal, otherwise the tests will default to the interactive authentication variant.

Sloane4 commented on PowerShell/ScubaGear/Testing/Unit/Rego/Sharepoint/SharepointConfig_03_test.rego:
```diff
@@ -153,73 +153,146 @@ test_SharingCapability_NewExistingGuests_NotApplicable_V2 if {
-                {"op": "add", "path": "FileAnonymousLinkType", "value": 2}])
+    # Set PnP flag to true denoting use of service principal
+    Output := sharepoint.tests with input.SPO_tenant as [Tenant]
+        with input.OneDrive_PnP_Flag as true
```
Isn't this supposed to be removed?

Sloane4 commented on PowerShell/ScubaGear/Testing/Unit/Rego/Sharepoint/SharepointConfig_03_test.rego:
```diff
@@ -153,73 +153,146 @@ test_SharingCapability_NewExistingGuests_NotApplicable_V2 if {
+    
+    # Set PnP flag to true denoting use of service principal 
+    Output := sharepoint.tests with input.SPO_tenant as [Tenant]
+        with input.OneDrive_PnP_Flag as true
```
Isn't this supposed to be removed?

Sloane4 commented on PowerShell/ScubaGear/Testing/Unit/Rego/Sharepoint/SharepointConfig_03_test.rego:
```diff
@@ -153,73 +153,146 @@ test_SharingCapability_NewExistingGuests_NotApplicable_V2 if {
+    
+    # Set PnP flag to true denoting use of service principal
+    Output := sharepoint.tests with input.SPO_tenant as [Tenant]
+        with input.OneDrive_PnP_Flag as true
```
Isn't this supposed to be removed?

tkol2022 commented:
@mitchelbaker-cisa Seems there is a logic flaw in the Rego implementation for policy 3.1 that I found when testing the branch. It is easy to fix so I recommend that you address it now especially given the importance of this release.

### To recreate the problem
Set the external sharing slider to Anyone

![image](https://github.com/user-attachments/assets/87ee0a9a-f8ff-41c4-a9ed-594dcf2ff820)

Uncheck the expiration days option
![image](https://github.com/user-attachments/assets/d0d57095-a53c-4a81-8449-0e0b398f8777)

Run ScubaGear and observe that it produces a Pass which is incorrect because the expiration option was not selected and configured
![image](https://github.com/user-attachments/assets/e4c52e24-b33e-48f3-9ba9-6fb8195bfa08)

### The fix
I believe the fix is relatively straightforward. When the expiration checkbox is unselected in the portal, you will notice that the JSON produces a value of -1 for RequireAnonymousLinksExpireInDays, so I am thinking that we need to ensure that the value of the field is between 1 and 30 inclusive. Currently the Rego code only checks for <= 30 and hence why it is flawed. Take a look and see if you agree.

![image](https://github.com/user-attachments/assets/b7d0b72a-2968-4b08-8cde-6f895a62eedd)

![image](https://github.com/user-attachments/assets/08ab2be6-2d90-4e99-8616-796fb6cede4f)

tkol2022 commented:
### N/A messages for section 3 are inconsistent with the baseline document
While testing I noticed that when policies 3.1, 3.2 or 3.3 produce an N/A, the messages provided in the report are inaccurate and/or do not align with the messages in the baseline document.

#### Policies 3.1 and 3.2 
Should say "This policy is only applicable if the external sharing slider on the admin center sharing page is set to Anyone." but in the report they say "This policy is only applicable if External Sharing is set to any value other than Anyone."
![image](https://github.com/user-attachments/assets/9e29f990-8dc5-4b2a-9372-c58c3db24d17)

#### Policy 3.3 
Should say "This policy is only applicable if the external sharing slider on the admin center sharing page is set to Anyone or New and existing guests." but in the report it says "This policy is only applicable if External Sharing is set to any value other than Only People In Your Organization or Existing Guests." In this case, technically the message in the report is correct but it does not align with the baseline so we might as well synch them up for consistency.

![image](https://github.com/user-attachments/assets/402114be-105a-4f0c-8690-293493d48454)

tkol2022 commented:
I tested interactive and non-interactive against E5 tenant and policy 3.2 seems to behave as expected.

mitchelbaker-cisa commented:
@tkol2022 

I addressed the logic for policy 3.1 to account for the negative integer value, -1, when the Anyone links checkbox is not set. The policy now displays a fail for this case.

The report details for policies 3.1-3.3 are also fixed.

tkol2022 commented:
> @tkol2022
> 
> I addressed the logic for policy 3.1 to account for the negative integer value, -1, when the Anyone links checkbox is not set. The policy now displays a fail for this case.
> 
> The report details for policies 3.1-3.3 are also fixed.

@Sloane4 Can you test the negative integer value logic and make sure it works for you? would be nice to have some independent validation.

Sloane4 reviewed (approved):
Dont forget to rebase with main

tkol2022 commented:
### Final Tests
I did a final functional test on all the Sharepoint policies that were affected by this PR 1.3, 3.1, 3.2, 3.3, 4.2 
They all passed except for 3.2. After further investigation the failures were intermittent and are not related to the code changes, but instead are related to the problems we have been experiencing with the Set-PnPTenant and Set-SPOTenant not updating the tenant in some cases. The problem seems to be random and we have informed Microsoft of the issue.
I also double checked and the unit tests are passing fine.
Also @ehaines1 Performed a hands-on test of the update to policy 3.1 based on the bug that was discovered during this PR review and it works as expected.

@nanda-katikaneni Ready for merge
